airbytehq/airbyte
# Fix issues with AvailabilityStrategy and substream stream_slices

- closes #20677 

## To reproduce the [KeyError issue in github](https://github.com/airbytehq/oncall/issues/1198) locally: 

Connect to the following Github repo (public, so any PAT should work)

<img width="658" alt="image (2)" src="https://user-images.githubusercontent.com/22333512/211367452-9c135f55-c549-40db-a111-893015d46717.png">

Turn on the `commits` and `commit_comment_reactions streams` with `Full refresh | Overwrite`. **Note**: you don't need the `commits` stream to reproduce, but I wanted to make sure that streams emitting records were still emitting records with the fix, and not just succeeding with 0 records.

Run a sync on a working version - `0.3.11` (latest) or or `0.3.8` (working version before broken version) should both work.
Run a sync on `0.3.9` (broken version) - sync will fail with the following traceback:
```
File "/airbyte/integration_code/main.py", line 13, in <module>
    launch(source, sys.argv[1:])
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py", line 131, in launch
    for message in source_entrypoint.run(parsed_args):
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/entrypoint.py", line 122, in run
    for message in generator:
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/abstract_source.py", line 109, in read
    stream_is_available, error = stream_instance.check_availability(logger, self)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/core.py", line 190, in check_availability
    return self.availability_strategy.check_availability(self, logger, source)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/availability_strategy.py", line 35, in check_availability
    stream_helper.get_first_record(stream)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/utils/stream_helpers.py", line 24, in get_first_record
    return next(records)
  File "/airbyte/integration_code/source_github/streams.py", line 939, in read_records
    for record in super().read_records(
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py", line 419, in read_records
    yield from self._read_pages(
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py", line 435, in _read_pages
    request, response = self._fetch_next_page(stream_slice, stream_state, next_page_token)
  File "/usr/local/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py", line 450, in _fetch_next_page
    path=self.path(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
  File "/airbyte/integration_code/source_github/streams.py", line 904, in path
    parent_path = self._parent_stream.path(stream_slice=stream_slice, **kwargs)
  File "/airbyte/integration_code/source_github/streams.py", line 492, in path
    return f"repos/{stream_slice['repository']}/comments"
KeyError: 'repository'
```

## To test the potential fix locally: 

Check out the branch, and add symlink to CDK:
```
cd airbyte-integrations/connectors/source-github
ln -s ../../../airbyte-cdk/python airbyte-cdk
```

Build the `source-github` image on the branch: 
```
cd ../../..
./gradlew airbyte-integrations:connectors:source-github:airbyteDocker
```

Temporary changes make the docker build reference the local CDK (which is branched off of re-adding in the AvailabilityStrategy to the CDK). This contains no changes to the Github code (i.e. no implementation of an AvailabilityStrategy other than the default HttpAvailabilityStrategy) but with the potential fix in the CDK version. Run sync with this newly created `dev` image. Should succeed, with the same 12 records.


## To reproduce the issue in SAT

Create a copy of `secrets/config.json` into `secrets/test_config.json`. Update the contents as so: 
```
{
  "credentials": {
    "personal_access_token": "<existing PAT, or your own>"
  },
  "repository": "erohmensing/thismonth.rocks",
  "start_date": "2015-01-01T00:00:00Z"
}
```
Now you should be able to run the SAT tests locally with the existing versions and `dev` version built with the CDK and get the same results as reproducing the issue above. E.g. `0.3.8` should pass, `0.3.9` should fail with the error, and `dev` should pass again. 

Note that [this line](https://github.com/airbytehq/airbyte/pull/20891/files#diff-bf8fabbee06c53040de42be93837cfea9bee5a2f0ee4cec087b435844264cfc7R15) was added because it would fail on returning no records (it shouldn't return any - there are none in the repo) in the working versions.
